Zoho Bookings API (v1) Integration

ysahil97 opened the pull request:
Hi Sam

As discussed, I have written some API's for Zoho Bookings API which we currently want to use.  The api's implemented here are related to appointments, staffs, resources, services and workspaces. Please let me know if there are any doubts or concerns in this pull request.

Commit:
Add zoho bookings scope support

Commit:
Add booking integration

Commit:
Implement all get api's of bookings

Commit:
Working implementation of post request in appointments

Commit:
Add readme for bookings

Commit:
# This is a combination of 2 commits.
# This is the 1st commit message:

minor line changes

# This is the commit message #2:

Some more minor issues

Commit:
revert back go.mod

schmorrison commented on bookings/appointment.go:
```diff
@@ -0,0 +1,155 @@
+	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
+}
+
+func (c *API) BookAppointment(request map[string]string) (data AppointmentResponse, err error) {
```
I'm still not sure how to pass this request type to the HTTPRequest method. I know I would like to see the function signature to either include the specific parameters, or a struct that defines the specific parameters, and use that as the `endpoint.RequestBody`.

To keep backwards compatible, and not introduce undefined behaviour, after reviewing the method I am thinking defining a new `endpoint.BodyFormat`, then you can specify the required body format in the endpoint.

If this pattern seems ok to you, we can discuss the exact changes to be made to HTTPRequest method, and how your new functions should be structured.

ysahil97 replied:
Yes, the pattern seems to be a good starting point.

schmorrison commented on bookings/appointment.go:
```diff
@@ -0,0 +1,155 @@
+	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
+}
+
+type UpdateAppointmentData struct {
+	BookingID string `json:"booking_id"`
+	Action string `json:"action"`
+}
+
+type RescheduleAppointmentData struct {
+	BookingID string `json:"booking_id"`
+	StaffId string `json:"staff_id,omitempty"`
+	StartTime string `json:"start_time,omitempty"`
+}
+type CustomerDetails struct {
+	Name string `json:"name"`
+	Email string `json:"email"`
+	PhoneNumber string `json:"phone_number"`
+}
+
+type BookAppointmentData struct {
+	ServiceId string `json:"service_id"`
+	StaffId string `json:"staff_id,omitempty"`
+	ResourceId string `json:"resource_id,omitempty"`
+	FromTime string `json:"from_time"`
+	TimeZone string `json:"time_zone,omitempty"`
+	Customer_Details CustomerDetails `json:"customer_details,omitempty"`
+}
```
These should be provided to the related methods instead of `map[string]string`

schmorrison commented on bookings/bookings.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+const (
+	BookingsAPIEndpoint string = "https://www.zohoapis.in/bookings/v1/json/"
```
View other subpackages to see how they handle the TLD.  Its very important because some users do not have access to the services behind some international TLDs.

schmorrison commented on bookings/resources.go:
```diff
@@ -0,0 +1,45 @@
+	zoho "github.com/schmorrison/Zoho"
+)
+
+func (c *API) FetchResources(request interface{}, params map[string]zoho.Parameter) (data ResourceResponse, err error) {
```
If I follow correctly, you are not providing anything to the `request` parameter (nil?) and using the `params` to provide the `resource_id` and `service_id`.

I would instead prefer to see the method signature to be the following:
```
func (c *API) FetchResources(resourceId string, serviceId string) (data ResourceResponse, err error)
```

and sending the params:
```
endpoint := zoho.Endpoint{
...
URLParameters: map[string]zoho.Parameter {
    "service_id": serviceId,
    "resource_id": resourceId,
    },
}
```

Use this pattern wherever possible.

schmorrison commented on oauth2.go:
```diff
@@ -268,7 +268,12 @@ type ScopeString string
 // BuildScope is used to generate a scope string for oAuth2 flow
 func BuildScope(service Service, scope Scope, method Method, operation Operation) ScopeString {
-	built := fmt.Sprintf("%s.%s.%s", service, scope, method)
+	var built string = ""
+	if method != "" {
+		built += fmt.Sprintf("%s.%s.%s", service, scope, method)
+	} else {
+		built += fmt.Sprintf("%s.%s", service, scope)
+	}
 	if operation != "" {
 		built += "." + string(operation)
 	}
```
Instead of creating an `else` line, See what we do with operation. This should be fine but could lead to weird errors if users are being silly.

```
built := fmt.Sprintf("%s.%s", service, scope)
if method != "" {
    built += fmt.Sprintf(".%s", method)
}
if operation != "" {
    built += fmt.Sprintf(".%s", operation)
}
```

ysahil97 replied:
Sure, the suggested code snippet makes more sense.

schmorrison commented on http.go:
```diff
@@ -68,15 +69,30 @@ func (z *Zoho) HTTPRequest(endpoint *Endpoint) (err error) {
+	if endpoint.RequestBody != nil{
+
+		// TODO: Decide whether JSON.marshal is required or not for
+		// Bookings or CRM
+
 		// JSON Marshal the body
-		marshalledBody, err := json.Marshal(endpoint.RequestBody)
-		if err != nil {
-			return fmt.Errorf("Failed to create json from request body")
+		switch v := endpoint.RequestBody.(type) {
+			case map[string]string:
+				body := url.Values{}
+				for k, val := range v {
+					body.Add(k,val)
+				}
+
+				reqBody = strings.NewReader(body.Encode())
+				contentType = "application/x-www-form-urlencoded; charset=UTF-8"
+			default:
+				marshalledBody, err := json.Marshal(v)
+				if err != nil {
+					return fmt.Errorf("Failed to create json from request body")
+				}
+
+				reqBody = bytes.NewBuffer(marshalledBody)
+				contentType = "application/x-www-form-urlencoded; charset=UTF-8"
 		}
```
This is a little fragile I expect. We will discuss how to structure this once we decide the best way to provide the proper formatting to the `requestBody`.

ysahil97 replied:
Sure, the primary motivation to these changes were to accomodate different variable types which could come up in `requestBody` parameter. At the moment, type switching made an apt sense for me. I am open to any other ideas regarding the structure used for different `requestBody` formats.

schmorrison commented on bookings/bookings.go:
```diff
@@ -0,0 +1,44 @@
+
+const (
+	BookingsAPIEndpoint string = "https://www.zohoapis.in/bookings/v1/json/"
+	GetAppointmentModule string = "getappointment"
+	FetchWorkspacesModule string = "workspaces"
+	FetchServicesModule string = "services"
+	FetchStaffModule string = "staffs"
+	FetchResourceModule string = "resources"
+	BookAppointmentModule string = "appointment"
+	RescheduleAppointmentModule string = "rescheduleappointment"
+	UpdateAppointmentModule string = "updateappointment"
```
I am not sure if we are using `strings` or have defined a type like `zoho.Module`. Check `crm` subpackage for reference.

ysahil97 replied:
I have checked the `crm` subpackage, in `crm.go`, `Module` is used which is a type alias for `string`

Commit:
Rectify minor issues

- Addressed TLD issue in bookings API
- In oauth2.go, handled scope building according to the suggestions
- In Bookings GET apis, url arguments are passed individually

ysahil97 committed:
Merge branch 'schmorrison:master' into book_api_post

Commit:
Add FetchAvailability API

Commit:
Merge branch 'book_api_post' of https://github.com/ysahil97/Zoho into book_api_post

schmorrison commented on bookings/staff.go:
```diff
@@ -0,0 +1,47 @@
+			"filter_by": "",
+		},
+	}
+	if serviceID != "" {
+		endpoint.URLParameters["service_id"] = serviceID
+	}
+	if staffID != "" {
+		endpoint.URLParameters["staff_id"] = staffID
+	}
```
Do you need to return an error here if those parameters are empty?

schmorrison replied:
I see it returns a list. NVM

schmorrison reviewed (commented):
I am going to try a couple options for changing the HTTPRequest method.

schmorrison commented:
Here is my view of things:

```github.com/schmorrison/Zoho/http.go

package zoho

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"reflect"
	"github.com/schmorrison/go-querystring/query"  // small change from 'github.com/google/go-querystring/query'
)

// Endpoint defines the data required to interact with most Zoho REST api endpoints
type Endpoint struct {
	Method        HTTPMethod
	URL           string
	Name          string
	ResponseData  interface{}
	RequestBody   interface{}
	URLParameters map[string]Parameter
	Headers       map[string]string
	BodyFormat    BodyFormat
	Attachment    string
}

// Parameter is used to provide URL Parameters to zoho endpoints
type Parameter string
type BodyFormat string

const (
	JSON        = ""
	JSON_STRING = "jsonString"
	FILE        = "file"
	URL			= "url" // Added new BodyFormat option
)

// HTTPRequest is the function which actually performs the request to a Zoho endpoint as specified by the provided endpoint
func (z *Zoho) HTTPRequest(endpoint *Endpoint) (err error) {
	if reflect.TypeOf(endpoint.ResponseData).Kind() != reflect.Ptr {
		return fmt.Errorf("Failed, you must pass a pointer in the ResponseData field of endpoint")
	}

	// Load and renew access token if expired
	err = z.CheckForSavedTokens()
	if err == ErrTokenExpired {
		err := z.RefreshTokenRequest()
		if err != nil {
			return fmt.Errorf("Failed to refresh the access token: %s: %s", endpoint.Name, err)
		}
	}

	// Retrieve URL parameters
	endpointURL := endpoint.URL
	q := url.Values{}
	for k, v := range endpoint.URLParameters {
		if v != "" {
			q.Set(k, string(v))
		}
	}

	var (
		req         *http.Request
		reqBody     io.Reader
		contentType string
	)

	// Has a body, likely a CRUD operation (still possibly JSONString)
	if endpoint.BodyFormat == JSON || endpoint.BodyFormat == JSON_STRING {
		if endpoint.RequestBody != nil {
			// JSON Marshal the body
			marshalledBody, err := json.Marshal(endpoint.RequestBody)
			if err != nil {
				return fmt.Errorf("Failed to create json from request body")
			}
			
			reqBody = bytes.NewReader(marshalledBody)
			contentType = "application/x-www-form-urlencoded; charset=UTF-8"
		}
	}

	if endpoint.BodyFormat == JSON_STRING || endpoint.BodyFormat == FILE {
		// Create a multipart form
		var b bytes.Buffer
		w := multipart.NewWriter(&b)

		switch endpoint.BodyFormat {
		case JSON_STRING:
			// Use the form to create the proper field
			fw, err := w.CreateFormField("JSONString")
			if err != nil {
				return err
			}
			// Copy the request body JSON into the field
			if _, err = io.Copy(fw, reqBody); err != nil {
				return err
			}

			// Close the multipart writer to set the terminating boundary
			err = w.Close()
			if err != nil {
				return err
			}

		case FILE:
			// Retreive the file contents
			fileReader, err := os.Open(endpoint.Attachment)
			if err != nil {
				return err
			}
			defer fileReader.Close()
			// Create the correct form field
			part, err := w.CreateFormFile("attachment", filepath.Base(endpoint.Attachment))
			if err != nil {
				return err
			}
			// copy the file contents to the form
			if _, err = io.Copy(part, fileReader); err != nil {
				return err
			}

			err = w.Close()
			if err != nil {
				return err
			}
		}

		reqBody = &b
		contentType = w.FormDataContentType()
	}

       // New BodyFormat encoding option
	if endpoint.BodyFormat == URL {
		body, err := query.Values(endpoint.RequestBody) // send struct into the newly imported package
		if err != nil {
			return err
		}

		reqBody = strings.NewReader(body.Encode()) // write to body
		contentType = "application/x-www-form-urlencoded; charset=UTF-8"
	}


      /// NOTHING CHANGED PAST THIS POINT

	req, err = http.NewRequest(string(endpoint.Method), fmt.Sprintf("%s?%s", endpointURL, q.Encode()), reqBody)
	if err != nil {
		return fmt.Errorf("Failed to create a request for %s: %s", endpoint.Name, err)
	}

	req.Header.Set("Content-Type", contentType)

	// Add global authorization header
	req.Header.Add("Authorization", "Zoho-oauthtoken "+z.oauth.token.AccessToken)

	// Add specific endpoint headers
	for k, v := range endpoint.Headers {
		req.Header.Add(k, v)
	}

	resp, err := z.client.Do(req)
	if err != nil {
		return fmt.Errorf("Failed to perform request for %s: %s", endpoint.Name, err)
	}

	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("Failed to read body of response for %s: got status %s: %s", endpoint.Name, resolveStatus(resp), err)
	}

	dataType := reflect.TypeOf(endpoint.ResponseData).Elem()
	data := reflect.New(dataType).Interface()

	err = json.Unmarshal(body, data)
	if err != nil {
		return fmt.Errorf("Failed to unmarshal data from response for %s: got status %s: %s", endpoint.Name, resolveStatus(resp), err)
	}

	endpoint.ResponseData = data

	return nil
}


..............................................................................
```



Please review this and let me know what you think. Test it out to make sure it works with your project and endpoints.

Thanks again for the PR!

schmorrison commented:
Forgot to show how to use the url encoding on the struct.

```github.com/schmorrison/Zoho/bookings/appointment.go
type CustomerDetails struct {
	Name string `json:"name"`
	Email string `json:"email"`
	PhoneNumber string `json:"phone_number"`
}

// Note the change to tag `url` instead of `json`
type BookAppointmentData struct {
	ServiceId string `url:"service_id"`
	StaffId string `url:"staff_id,omitempty"`
	ResourceId string `url:"resource_id,omitempty"`
	FromTime string `url:"from_time"`
	TimeZone string `url:"time_zone,omitempty"`
	Customer_Details CustomerDetails `url:"customer_details,json,omitempty"` // Note the option `json` before `omitempty`, the order shouldn't matter
}

// The struct as the given request body
func (c *API) BookAppointment(request BookAppointmentData) (data AppointmentResponse, err error) {
	endpoint := zoho.Endpoint{
		Name:         BookAppointmentModule,
		URL:          fmt.Sprintf("https://www.zohoapis.%s/bookings/v1/json/%s",c.ZohoTLD,BookAppointmentModule),
		Method:       zoho.HTTPPost,
		ResponseData: &AppointmentResponse{},
		RequestBody: request,
                BodyFormat: zoho.URL, // Use the URL BodyFormat
	}

	err = c.Zoho.HTTPRequest(&endpoint)
	if err != nil {
		return AppointmentResponse{}, fmt.Errorf("Failed to book appointment: %s", err)
	}
	if v, ok := endpoint.ResponseData.(*AppointmentResponse); ok {

		return *v, nil
	}
	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
}
```

ysahil97 commented:
Hi Sam

The above suggested workaround to accommodate Zoho Booking requests seems
to be good for me.
I have tested with all the post API's present in Zoho Bookings API, and
they seem to work fine.

I have an additional point to make towards one of the API's, namely
BookAppointment api.
In this API, we also need to enter some more personal information but
customer_details only provides fields for name, email and phone number.
As a workaround, Zoho Bookings has created a custom field API (link) which
would solve this problem.
The issue with incorporating custom fields in our PR is that, because of
their nature of the fields, these are created separately for each project
case and therefore there can't be any standardized library implementation
of it.
For our use case, we are planning to maintain a fork of your library and we
would be using that fork as the library for our use cases. I will also add
an README note for the same.

Could we finalize on your changes and bring closure to this PR?
PS: In your fork of go-querystring, the go.mod file needs to be changed to
reflect proper owner name.

Best Regards
Sahil Yerawar

On Sun, Jul 25, 2021 at 12:11 AM Sam Morrison ***@***.***>
wrote:

> Forgot to show how to use the url encoding on the struct.
>
> type CustomerDetails struct {
> 	Name string `json:"name"`
> 	Email string `json:"email"`
> 	PhoneNumber string `json:"phone_number"`
> }
>
> // Note the change to tag `url` instead of `json`
> type BookAppointmentData struct {
> 	ServiceId string `url:"service_id"`
> 	StaffId string `url:"staff_id,omitempty"`
> 	ResourceId string `url:"resource_id,omitempty"`
> 	FromTime string `url:"from_time"`
> 	TimeZone string `url:"time_zone,omitempty"`
> 	Customer_Details CustomerDetails `url:"customer_details,json,omitempty"` // Note the option `json` before `omitempty`, the order shouldn't matter
> }
>
> // The struct as the given request body
> func (c *API) BookAppointment(request BookAppointmentData) (data AppointmentResponse, err error) {
> 	endpoint := zoho.Endpoint{
> 		Name:         BookAppointmentModule,
> 		URL:          fmt.Sprintf("https://www.zohoapis.%s/bookings/v1/json/%s",c.ZohoTLD,BookAppointmentModule),
> 		Method:       zoho.HTTPPost,
> 		ResponseData: &AppointmentResponse{},
> 		RequestBody: request,
>                 BodyFormat: zoho.URL, // Use the URL BodyFormat
> 	}
>
> 	err = c.Zoho.HTTPRequest(&endpoint)
> 	if err != nil {
> 		return AppointmentResponse{}, fmt.Errorf("Failed to book appointment: %s", err)
> 	}
> 	if v, ok := endpoint.ResponseData.(*AppointmentResponse); ok {
>
> 		return *v, nil
> 	}
> 	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/schmorrison/Zoho/pull/28#issuecomment-886095062>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEWHICSPCSILPKZ445ZCTB3TZMCPFANCNFSM476O4GFA>
> .
>


-- 
Thanking You
Sahil Yerawar

Commit:
Resolve Zoho Booking Body Format

ysahil97 commented:
Sorry, I forgot to add the link to the custom field API documentation:
https://www.zoho.com/bookings/help/api/v1/book-appointment.html

On Mon, Jul 26, 2021 at 4:11 PM Sahil Yerawar ***@***.***>
wrote:

> Hi Sam
>
> The above suggested workaround to accommodate Zoho Booking requests seems
> to be good for me.
> I have tested with all the post API's present in Zoho Bookings API, and
> they seem to work fine.
>
> I have an additional point to make towards one of the API's, namely
> BookAppointment api.
> In this API, we also need to enter some more personal information but
> customer_details only provides fields for name, email and phone number.
> As a workaround, Zoho Bookings has created a custom field API (link) which
> would solve this problem.
> The issue with incorporating custom fields in our PR is that, because of
> their nature of the fields, these are created separately for each project
> case and therefore there can't be any standardized library implementation
> of it.
> For our use case, we are planning to maintain a fork of your library and
> we would be using that fork as the library for our use cases. I will also
> add an README note for the same.
>
> Could we finalize on your changes and bring closure to this PR?
> PS: In your fork of go-querystring, the go.mod file needs to be changed to
> reflect proper owner name.
>
> Best Regards
> Sahil Yerawar
>
> On Sun, Jul 25, 2021 at 12:11 AM Sam Morrison ***@***.***>
> wrote:
>
>> Forgot to show how to use the url encoding on the struct.
>>
>> type CustomerDetails struct {
>> 	Name string `json:"name"`
>> 	Email string `json:"email"`
>> 	PhoneNumber string `json:"phone_number"`
>> }
>>
>> // Note the change to tag `url` instead of `json`
>> type BookAppointmentData struct {
>> 	ServiceId string `url:"service_id"`
>> 	StaffId string `url:"staff_id,omitempty"`
>> 	ResourceId string `url:"resource_id,omitempty"`
>> 	FromTime string `url:"from_time"`
>> 	TimeZone string `url:"time_zone,omitempty"`
>> 	Customer_Details CustomerDetails `url:"customer_details,json,omitempty"` // Note the option `json` before `omitempty`, the order shouldn't matter
>> }
>>
>> // The struct as the given request body
>> func (c *API) BookAppointment(request BookAppointmentData) (data AppointmentResponse, err error) {
>> 	endpoint := zoho.Endpoint{
>> 		Name:         BookAppointmentModule,
>> 		URL:          fmt.Sprintf("https://www.zohoapis.%s/bookings/v1/json/%s",c.ZohoTLD,BookAppointmentModule),
>> 		Method:       zoho.HTTPPost,
>> 		ResponseData: &AppointmentResponse{},
>> 		RequestBody: request,
>>                 BodyFormat: zoho.URL, // Use the URL BodyFormat
>> 	}
>>
>> 	err = c.Zoho.HTTPRequest(&endpoint)
>> 	if err != nil {
>> 		return AppointmentResponse{}, fmt.Errorf("Failed to book appointment: %s", err)
>> 	}
>> 	if v, ok := endpoint.ResponseData.(*AppointmentResponse); ok {
>>
>> 		return *v, nil
>> 	}
>> 	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
>> }
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/schmorrison/Zoho/pull/28#issuecomment-886095062>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEWHICSPCSILPKZ445ZCTB3TZMCPFANCNFSM476O4GFA>
>> .
>>
>
>
> --
> Thanking You
> Sahil Yerawar
>


-- 
Thanking You
Sahil Yerawar

schmorrison commented:
I can't review/merge right now (ill try during this week), but take a look at this with regard to custom fields, struct embedding is almost exactly that. Its been on the roadmap forever.

https://play.golang.org/p/pH4iB-6VJ-a

It does require a bit of looking at, most Zoho responses come wrapped in something, for bookings you get `{"response":{"returnValue": {......}, logMessage: [], status: ""}}` and we would really want to do the embedding at the `returnValue` level. There are a couple ways to reconfigure the structs to allow it, but right now most structs in most packages are just faithful conversions from JSON.

If you create an issue about this we can decide the best way in a separate thread.

schmorrison commented on bookings/bookings.go:
```diff
@@ -0,0 +1,44 @@
+	zoho "github.com/schmorrison/Zoho"
+	"math/rand"
+)
+
```
```go
type BookingsModule = string
```

schmorrison commented on bookings/bookings.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+const (
+	GetAppointmentModule string = "getappointment"
+	GetAvailabilityModule string = "availableslots"
+	FetchWorkspacesModule string = "workspaces"
+	FetchServicesModule string = "services"
+	FetchStaffModule string = "staffs"
+	FetchResourceModule string = "resources"
+	BookAppointmentModule string = "appointment"
+	RescheduleAppointmentModule string = "rescheduleappointment"
+	UpdateAppointmentModule string = "updateappointment"
```
```go
	GetAppointmentModule BookingsModule = "getappointment"
	GetAvailabilityModule BookingsModule = "availableslots"
	FetchWorkspacesModule BookingsModule = "workspaces"
	FetchServicesModule BookingsModule = "services"
	FetchStaffModule BookingsModule  = "staffs"
	FetchResourceModule BookingsModule = "resources"
	BookAppointmentModule BookingsModule = "appointment"
	RescheduleAppointmentModule BookingsModule = "rescheduleappointment"
	UpdateAppointmentModule BookingsModule = "updateappointment"
```

schmorrison commented on bookings/appointment.go:
```diff
@@ -0,0 +1,160 @@
+	return AppointmentResponse{}, fmt.Errorf("Data retrieved was not 'AppointmentResponse'")
+}
+
+func (c *API) RescheduleAppointment(params RescheduleAppointmentData) (data AppointmentResponse, err error) {
```
```go 
func (c *API) RescheduleAppointment(request RescheduleAppointmentData) (data AppointmentResponse, err error)
```

schmorrison commented on bookings/staff.go:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	if v,ok := endpoint.ResponseData.(*StaffResponse); ok {
+		fmt.Printf("okay")
```
remove debug print

schmorrison commented on go.mod:
```diff
@@ -2,4 +2,7 @@ module github.com/schmorrison/Zoho
 
-require google.golang.org/appengine v1.6.6
+require (
+	github.com/ysahil97/go-querystring v1.1.1-0.20210726100856-5f3003a0fbff // indirect
```
```
github.com/schmorrison/go-querystring v1.1.1
```

schmorrison commented on http.go:
```diff
@@ -12,6 +12,8 @@ import (
 	"path/filepath"
 	"reflect"
+	"strings"
+	"github.com/ysahil97/go-querystring/query"
```
```
github.com/schmorrison/go-querystring/query
```

schmorrison commented on http.go:
```diff
@@ -68,70 +71,107 @@ func (z *Zoho) HTTPRequest(endpoint *Endpoint) (err error) {
-			err = w.Close()
-			if err != nil {
-				return err
+	// if endpoint.RequestBody != nil{
+	//
+	// 	// TODO: Decide whether JSON.marshal is required or not for
+	// 	// Bookings or CRM
+	//
+	// 	// JSON Marshal the body
+	// 	switch v := endpoint.RequestBody.(type) {
+	// 	case url.Values:
+	// 			// body := url.Values{}
+	// 			// for k, val := range v {
+	// 			// 	body.Add(k,val)
+	// 			// }
+	//
+	// 			reqBody = strings.NewReader(v.Encode())
+	// 			fmt.Println(reqBody)
+	// 			contentType = "application/x-www-form-urlencoded; charset=UTF-8"
+	// 		default:
+	// 			marshalledBody, err := json.Marshal(v)
+	// 			if err != nil {
+	// 				return fmt.Errorf("Failed to create json from request body")
+	// 			}
+	//
+	// 			reqBody = bytes.NewBuffer(marshalledBody)
+	// 			contentType = "application/x-www-form-urlencoded; charset=UTF-8"
+	// 	}
+	// }
+
```
remove comment

schmorrison commented on bookings/appointment.go:
```diff
@@ -0,0 +1,160 @@
+
+import (
+	"fmt"
+	// "net/url"
```
remove import

schmorrison reviewed (changes requested):
Finally run ```go fmt```.

Thanks for your contributions and patience!!

Commit:
Final changes for the Bookings PR

Commit:
Remove dead imports

schmorrison reviewed (approved):
Thanks @ysahil97 for you contribution!

ysahil97 commented:
Your welcome!
Could you please merge the changes in the main branch?
